File: frontend/serve.js
#!/usr/bin/env node
/**
 * Simple Node.js HTTP Server for Septica PWA Testing
 * Alternative to Python server with same functionality
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

const PORT = process.env.PORT || 8000;
const MIME_TYPES = {
    '.html': 'text/html',
    '.css': 'text/css',
    '.js': 'application/javascript',
    '.json': 'application/json',
    '.webmanifest': 'application/manifest+json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon',
    '.woff': 'font/woff',
    '.woff2': 'font/woff2'
};

function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    return MIME_TYPES[ext] || 'application/octet-stream';
}

function serveFile(res, filePath) {
    fs.readFile(filePath, (err, data) => {
        if (err) {
            res.writeHead(404, { 'Content-Type': 'text/html' });
            res.end('<h1>404 Not Found</h1>');
            return;
        }
        
        const mimeType = getMimeType(filePath);
        const headers = {
            'Content-Type': mimeType,
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type'
        };
        
        // Service worker should not be cached
        if (path.basename(filePath) === 'sw.js') {
            headers['Cache-Control'] = 'no-cache, no-store, must-revalidate';
            headers['Pragma'] = 'no-cache';
            headers['Expires'] = '0';
        }
        
        res.writeHead(200, headers);
        res.end(data);
    });
}

const server = http.createServer((req, res) => {
    const parsedUrl = url.parse(req.url);
    let pathname = parsedUrl.pathname;
    
    // Handle root path
    if (pathname === '/') {
        pathname = '/index.html';
    }
    
    // Security: prevent directory traversal
    const safePath = path.normalize(pathname).replace(/^(\.\.[\/\\])+/, '');
    const filePath = path.join(__dirname, safePath);
    
    // Check if file exists
    fs.stat(filePath, (err, stats) => {
        if (err || !stats.isFile()) {
            // Try to serve index.html for PWA routing
            const indexPath = path.join(__dirname, 'index.html');
            fs.stat(indexPath, (indexErr) => {
                if (indexErr) {
                    res.writeHead(404, { 'Content-Type': 'text/html' });
                    res.end('<h1>404 Not Found</h1>');
                } else {
                    serveFile(res, indexPath);
                }
            });
        } else {
            serveFile(res, filePath);
        }
    });
});

server.listen(PORT, () => {
    console.log('🎮 Septica PWA Server');
    console.log(`📡 Serving at http://localhost:${PORT}`);
    console.log(`📁 Serving directory: ${__dirname}`);
    console.log(`🔗 Open http://localhost:${PORT} in your browser`);
    console.log(`⚡ Make sure your Go backend is running on ws://localhost:8080/ws/connect`);
    console.log(`🛑 Press Ctrl+C to stop`);
    console.log();
});

server.on('error', (err) => {
    if (err.code === 'EADDRINUSE') {
        console.log(`❌ Port ${PORT} is already in use`);
        console.log(`💡 Try using a different port: PORT=8001 node serve.js`);
    } else {
        console.log(`❌ Error starting server: ${err.message}`);
    }
    process.exit(1);
});

process.on('SIGINT', () => {
    console.log('\n👋 Server stopped');
    process.exit(0);
});

File: frontend/serve.py
#!/usr/bin/env python3
"""
Simple HTTP Server for Septica PWA Testing
Serves the PWA with proper MIME types and CORS headers
"""

import http.server
import socketserver
import os
import sys
from pathlib import Path

# Set the port
PORT = 8000

# Change to the directory containing this script
os.chdir(Path(__file__).parent)

class PWAHandler(http.server.SimpleHTTPRequestHandler):
    """Custom handler with proper MIME types and headers for PWA"""
    
    def end_headers(self):
        # Add CORS headers for development
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        
        # Add PWA-specific headers
        if self.path.endswith('.js'):
            self.send_header('Content-Type', 'application/javascript')
        elif self.path.endswith('.json'):
            self.send_header('Content-Type', 'application/json')
        elif self.path.endswith('.css'):
            self.send_header('Content-Type', 'text/css')
        elif self.path.endswith('.html'):
            self.send_header('Content-Type', 'text/html')
        
        # Service worker should not be cached
        if self.path.endswith('sw.js'):
            self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
            self.send_header('Pragma', 'no-cache')
            self.send_header('Expires', '0')
        
        super().end_headers()
    
    def guess_type(self, path):
        """Override to handle additional file types"""
        mime_type = super().guess_type(path)
        if path.endswith('.js'):
            return 'application/javascript'
        elif path.endswith('.json'):
            return 'application/json'
        elif path.endswith('.webmanifest'):
            return 'application/manifest+json'
        return mime_type

def main():
    try:
        with socketserver.TCPServer(("", PORT), PWAHandler) as httpd:
            print(f"🎮 Septica PWA Server")
            print(f"📡 Serving at http://localhost:{PORT}")
            print(f"📁 Serving directory: {os.getcwd()}")
            print(f"🔗 Open http://localhost:{PORT} in your browser")
            print(f"⚡ Make sure your Go backend is running on ws://localhost:8080/ws/connect")
            print(f"🛑 Press Ctrl+C to stop")
            print()
            
            httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n👋 Server stopped")
        sys.exit(0)
    except OSError as e:
        if e.errno == 48:  # Address already in use
            print(f"❌ Port {PORT} is already in use")
            print(f"💡 Try using a different port or stop the other server")
        else:
            print(f"❌ Error starting server: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()